moleculemaker/mmli-job-manager
# Implement authentication via oauth2 proxy

## Problem
We need the JobManager API to include a small bit of Python to handle authentication/authorization

## Approach
* Deployed OAuth2 Proxy + Keycloak to the cluster: https://mmli1.nsca.illinois.edu/oauth2/start?rd=/oauth2/userinfo
* Cookie set by OAuth2 Proxy will be sent to JobManager API (if we set the cookie on the correct domains)
* Python checks for this cookie, and sends it back to OAuth2 Proxy to validate - this is similar to what we currently do for HCAPTCHA
    * If user is returned from our request, then we have our username, email, groups, etc
    * If no user is returned, then we return 401: Unauthorized
* Allow admin to override configs at runtime (this is still a little too fragile for my liking)

## How to Test
1. (already running in `staging`) Start this image up on the cluster using ArgoCD: `moleculemaker/mmli-job-manager:pr-22`
2. Navigate to https://jobmgr.mmli1.ncsa.illinois.edu/api/v1/uws/job
    * You should receive an error saying `401: Unauthorized` 
3. Navigate to https://mmli1.ncsa.illinois.edu/oauth2/start?rd=https://jobmgr.staging.mmli1.ncsa.illinois.edu/api/v1/uws/job
    * You should be redirected to Keycloak to login
4. Login or create an account in Keycloak
    * You should be redirected back to the JobManager API after you log in
    * You should see the list of Jobs submitted here
5. Logout by navigating to https://mmli1.ncsa.illinois.edu/oauth2/sign_out?rd=https://jobmgr.staging.mmli1.ncsa.illinois.edu/api/v1/uws/job
    * You should be redirected back to the JobManager API after you log out
    * You should once again see an error saying `401: Unauthorized`
